src/common/hist_util.h

```diff
@@ -190,6 +190,101 @@ size_t DeviceSketch(int device,
                     DMatrix* dmat,
                     HistogramCuts* hmat);
 
+
+enum BinBounds {
+  UINT8_BINS_TYPE = 0,
+  UINT16_BINS_TYPE,
+  UINT32_BINS_TYPE,
+};
+
+struct Index {
+  Index(): binBound_(UINT8_BINS_TYPE), p_(1), disp_ptr_(nullptr) {
+    setBinBound(binBound_);
+  }
+  Index(const Index& i) = delete;
+  Index& operator=(Index i) = delete;
+  Index(Index&& i) = delete;
+  Index& operator=(Index&& i) = delete;
+  uint32_t operator[](size_t i) const {
+    if (disp_ptr_ != nullptr) {
```

Comment:
When is this pointer being null?

Reply:
For sparse DMarix this `offset_ptr_` field (in previous commit 'disp_ptr_') is `nullptr`, reasons were described above:
https://github.com/dmlc/xgboost/pull/5334#discussion_r395643869